src/server.coffee

```diff
@@ -44,438 +51,526 @@ tcpAdapter = require './tcpAdapter'
-  return defer.promise
-
-startHttpServer = (httpPort, bindAddress, app) ->
-  deferred = Q.defer()
-
-  httpServer = http.createServer app.callback()
-
-  # set the socket timeout
-  httpServer.setTimeout config.router.timeout, ->
-    logger.info "HTTP socket timeout reached"
-
-  httpServer.listen httpPort, bindAddress, ->
-    logger.info "HTTP listening on port " + httpPort
-    deferred.resolve()
-
-  httpServer.on 'connection', (socket) -> trackConnection activeHttpConnections, socket
-
-  return deferred.promise
-
-startHttpsServer = (httpsPort, bindAddress, app) ->
-  deferred = Q.defer()
-
-  mutualTLS = config.authentication.enableMutualTLSAuthentication
-  tlsAuthentication.getServerOptions mutualTLS, (err, options) ->
-    return done err if err
-    httpsServer = https.createServer options, app.callback()
+clusterArg = nconf.get 'cluster'
 
-    # set the socket timeout
-    httpsServer.setTimeout config.router.timeout, ->
-      logger.info "HTTPS socket timeout reached"
+# Configure clustering if relevent
+if cluster.isMaster and not module.parent
 
-    httpsServer.listen httpsPort, bindAddress, ->
-      logger.info "HTTPS listening on port " + httpsPort
-      deferred.resolve()
+  # configure master logger
+  logger.add logger.transports.Console,
+    colorize: true
+    label: "master"
+    level: config.logger.level
 
-    httpsServer.on 'connection', (socket) -> trackConnection activeHttpsConnections, socket
+  if not clusterArg?
+    clusterArg = 1
 
-  return deferred.promise
+  if clusterArg is 'auto'
+    clusterSize = numCPUs
+  else
+    clusterSize = clusterArg
+
+  if typeof clusterSize isnt 'number' or clusterSize % 1 isnt 0 or clusterSize < 1
+    throw new Error "invalid --cluster argument entered: #{clusterArg}. Please enter a positive number or 'auto'."
+
+  logger.info "Clustering the OpenHIM with #{clusterSize} workers..."
+
+  addWorker = () ->
+    worker = cluster.fork()
+
+    worker.on 'message', (msg) ->
+      if msg is 'restart-all'
+        # restart all workers
+        logger.debug "Restarting all workers..."
+        for id, worker of cluster.workers
+          logger.debug "Restarting worker #{worker.id}..."
+          worker.send 'restart'
+
+  # start all workers
+  for i in [1..clusterSize]
+    addWorker()
+
+  cluster.on 'exit', (worker, code, signal) ->
+    logger.warn "worker #{worker.process.pid} died"
+    if not worker.suicide
+      # respawn
+      addWorker()
+
+  cluster.on 'online', (worker) ->
+    logger.info "worker #{worker.process.pid} is online"
+
+  cluster.on 'listening', (worker, address) ->
+    logger.debug "worker #{worker.process.pid} is now connected to #{address.address}:#{address.port}"
+else
+  # configure worker logger
+  logger.add logger.transports.Console,
+    colorize: true
+    label: "worker#{cluster.worker.id}" if cluster.worker?.id?
+    level: config.logger.level
+
+  # Configure mongose to connect to mongo
+  mongoose.connect config.mongo.url
```

Comment:
Out of scope question :) Is this still needed with the createConnection calls above?